### PR TITLE
li2_nc reader daskified

### LINKED ISSUE #2814: Reading LI L2 point data is not daskified
**Describe the bug**

Loading "radiance" from a LI L2 LGR file gives me the data directly, instead of a dask-aware array.

**To Reproduce**

```python
# Your code here
from satpy import Scene
from satpy.utils import debug_on; debug_on()
from glob import glob
sc = Scene(filenames=glob("/media/nas/x21308/MTG_test_data/LI/20230712/*LGR*_0042_003*.nc"), reader="li_l2_nc")
sc.load(["radiance"])
print(sc["radiance"])
```

**Expected behavior**

I expect that the NetCDF dataset encapsulates a dask array.

**Actual results**

It encapsulates a regular array, with data values, as follows:

```
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.yaml_reader] Reading ('/data/gholl/checkouts/satpy/satpy/etc/readers/li_l2_nc.yaml',)                                                                                                    [5/1066]
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.yaml_reader] Assigning to li_l2_nc: ['/media/nas/x21308/MTG_test_data/LI/20230712/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023504_L2PF_IV_20230713065623
_20230713065633_N__T_0042_0039.nc', '/media/nas/x21308/MTG_test_data/LI/20230712/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023408_L2PF_IV_20230713065513_20230713065523_N__T_0042_0032.nc', '/media/n
as/x21308/MTG_test_data/LI/20230712/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023432_L2PF_IV_20230713065543_20230713065553_N__T_0042_0035.nc', '/media/nas/x21308/MTG_test_data/LI/20230712/W_XX-EUME
TSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023424_L2PF_IV_20230713065533_20230713065543_N__T_0042_0034.nc', '/media/nas/x21308/MTG_test_data/LI/20230712/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK
-BODY---NC4E_C_EUMT_20231221023448_L2PF_IV_20230713065603_20230713065613_N__T_0042_0037.nc', '/media/nas/x21308/MTG_test_data/LI/20230712/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023456_L2PF_IV_20
230713065613_20230713065623_N__T_0042_0038.nc', '/media/nas/x21308/MTG_test_data/LI/20230712/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023440_L2PF_IV_20230713065553_20230713065603_N__T_0042_0036.nc
', '/media/nas/x21308/MTG_test_data/LI/20230712/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023352_L2PF_IV_20230713065453_20230713065503_N__T_0042_0030.nc', '/media/nas/x21308/MTG_test_data/LI/202307
12/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023416_L2PF_IV_20230713065523_20230713065533_N__T_0042_0033.nc', '/media/nas/x21308/MTG_test_data/LI/20230712/W_XX-EUMETSAT-Darmstadt,IMG+SAT,MTI1+LI-2-
LGR--FD--CHK-BODY---NC4E_C_EUMT_20231221023400_L2PF_IV_20230713065503_20230713065513_N__T_0042_0031.nc']
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Product type is: 2-LGR
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_base_nc] Adding 10 datasets for 2-LGR input product.
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.li_l2_nc] The current product 2-LGR is not an accumulated product so it will not be regridded.
[DEBUG: 2024-06-13 12:47:02 : satpy.composites.config_loader] Looking for composites config file li.yaml
[DEBUG: 2024-06-13 12:47:02 : satpy.composites.config_loader] Looking for composites config file visir.yaml
[DEBUG: 2024-06-13 12:47:02 : pyorbital.tlefile] Path to the Pyorbital configuration (where e.g. platforms.txt is found): /data/gholl/mambaforge/envs/py312/lib/python3.12/site-packages/pyorbital/etc
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.yaml_reader] No coordinates found for DataID(name='longitude', modifiers=())
[DEBUG: 2024-06-13 12:47:02 : satpy.readers.yaml_reader] No coordinates found for DataID(name='latitude', modifiers=())
<xarray.DataArray 'radiance' (y: 7105)> Size: 28kB
array([ 6. ,  4. , 13. , ...,  8. ,  6.5, 22. ], dtype=float32)
Coordinates:
    crs      object 8B +proj=longlat +ellps=WGS84 +type=crs
Dimensions without coordinates: y
Attributes: (12/18)
    long_name:            Radiance of group
    units:                mW.m-2.sr-1
    use_rescaling:        True
    sensor:               li
    add_offset:           0.0
    name:                 radiance
    ...                   ...
    start_time:           2023-07-13 06:54:53
    end_time:             2023-07-13 06:56:33
    reader:               li_l2_nc
    area:                 Shape: (7105,)\nLons: <xarray.DataArray 'longitude'...
    _satpy_id:            DataID(name='radiance', modifiers=())
    ancillary_variables:  []
```

**Environment Info:**
 - OS: openSUSE Leap 15.3
 - Satpy Version: main branch v0.49.0-55-g367016e29

**Additional context**

I encapsulated the whole thing in `with dask.config.set(scheduler=CustomScheduler(max_computes=0))`, expecting it to fail, but it doesn't fail.  So it looks like nothing is being computed, it just never is a dask array to begin with.

Reading gridded products, such as AFA, correctly results in a dask array.

### PR DESCRIPTION
**Pull request by ClementLaplace**
This pull request enables the daskification of the non accumulated and non transformed dataset. 

 - [x] Closes #2814
 - [x] Test that all the dataset are daskified for the the li_l2_nc reader

### CODE REVIEW

File: satpy/readers/li_l2_nc.py [hunk 1]
```diff
@@ -111,6 +111,10 @@ def get_dataset(self, dataset_id, ds_info=None):
         var_with_swath_coord = self.is_var_with_swath_coord(dataset_id)
         if var_with_swath_coord and self.with_area_def:
             data_array = self.get_array_on_fci_grid(data_array)
+        else :
+            if data_array is not None:
+                if not isinstance(data_array.data,da.Array):
```
Comment (mraspaud) [inline satpy/readers/li_l2_nc.py:116]:
```suggestion
                if not isinstance(data_array.data, da.Array):

```

File: satpy/tests/reader_tests/test_li_l2_nc.py [hunk 2]
```diff
@@ -128,6 +129,7 @@ def _test_dataset_variable(self, var_params, sname=""):
         res = self.get_variable_dataset(dataset_info, dname, handler)
         assert res.shape == shape
         assert res.dims[0] == "y"
+        assert isinstance(res.data,da.Array)
```
Comment (mraspaud) [inline satpy/tests/reader_tests/test_li_l2_nc.py:132]: was this failing before the other changes were implemented?

**Review by mraspaud — APPROVED** [review]
LGTM, just a comment or two inline.

File: satpy/tests/reader_tests/test_li_l2_nc.py (hunk 2 above)
Comment (ClementLaplace) [reply satpy/tests/reader_tests/test_li_l2_nc.py:132]: I have put in commentary my implementation and run the tests and it is failing as expected 
```python 
test_li_l2_nc.py F...............................                                                                                                                                                      [100%]

================================================================================================== FAILURES ==================================================================================================
_______________________________________________________________________________________ TestLIL2.test_dataset_loading ________________________________________________________________________________________

self = <satpy.tests.reader_tests.test_li_l2_nc.TestLIL2 object at 0x7efc7f2da310>
filetype_infos = {'li_l2_af_nc': {'file_desc': {'product_type': '2-AF', 'swath_coordinates': {'azimuth': 'x', 'elevation': 'y', 'projec..._cycle:>04d}.nc'], 'file_reader': <class 'satpy.readers.li_l2_nc.LIL2NCFileHandler'>, 'file_type': 'li_l2_le_nc'}, ...}

    def test_dataset_loading(self, filetype_infos):
        """Test loading of all datasets from all products."""
        # Iterate on all the available product types:
        for ptype, pinfo in products_dict.items():
            ftype = pinfo["ftype"]
            filename_info = {
                "start_time": "0000",
                "end_time": "1000"
            }

            handler = LIL2NCFileHandler("filename", filename_info, extract_filetype_info(filetype_infos, ftype),
                                        with_area_definition=False)
            ds_desc = handler.ds_desc

            # retrieve the schema that what used to generate the content for that product:
            settings = get_product_schema(ptype)

            # Now we check all the variables are available:
            if "variables" in settings:
                self._test_dataset_variables(settings, ds_desc, handler)

                # check the sector variables:
            if "sector_variables" in settings:
>               self._test_dataset_sector_variables(settings, ds_desc, handler)

test_li_l2_nc.py:194:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_li_l2_nc.py:169: in _test_dataset_sector_variables
    self._test_dataset_single_sector_variable([sname, vname], desc, settings, handler)
test_li_l2_nc.py:124: in _test_dataset_single_sector_variable
    self._test_dataset_variable(var_params, sname=sname)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <satpy.tests.reader_tests.test_li_l2_nc.TestLIL2 object at 0x7efc7f2da310>
var_params = [{'name': 'event_id_north_sector', 'sector_name': 'north', 'use_rescaling': False, 'variable_name': 'event_id'}, {'def..., 'format': 'u2', 'long_name': 'ID of LI L2 Event', ...}, 'event_id', <LIL2NCFileHandler: 'filename'>, (123,), 'data/']
sname = 'north'

    def _test_dataset_variable(self, var_params, sname=""):
        """Test the validity of a given (sector) variable."""
        dataset_info, desc, dname, handler, shape, var_path = var_params
        res = self.get_variable_dataset(dataset_info, dname, handler)
        assert res.shape == shape
        assert res.dims[0] == "y"
>       assert isinstance(res.data,da.Array)
E       AssertionError: assert False
E        +  where False = isinstance(array([43325., 31613., 47855., 64957., 29417.,  5231., 23702.,  2226.,\n        1884., 21431., 29262., 52112.,  2193., ...,\n       16686., 48404., 60604., 54068., 19620., 55670., 27682., 46129.,\n       37781., 44337.,  2549.], dtype=float32), <class 'dask.array.core.Array'>)
E        +    where array([43325., 31613., 47855., 64957., 29417.,  5231., 23702.,  2226.,\n        1884., 21431., 29262., 52112.,  2193., ...,\n       16686., 48404., 60604., 54068., 19620., 55670., 27682., 46129.,\n       37781., 44337.,  2549.], dtype=float32) = <xarray.DataArray (y: 123)> Size: 492B\narray([43325., 31613., 47855., 64957., 29417.,  5231., 23702.,  2226.,\n        ... name:           event_id_north_sector\n    variable_name:  event_id\n    sector_name:    north\n    use_rescaling:  False.data
E        +    and   <class 'dask.array.core.Array'> = da.Array

test_li_l2_nc.py:132: AssertionError
```

**Comment by ameraner** [comment]
So this PR is checking if an array is not daskified and converts it to a dask array if it's not. 

I think it would be important to know why the array was not a dask array in the first place, otherwise we are still possibly loading the data into memory maybe unnecessarily. Tracking back the `get_dataset` in the base class, the variables are taken from the file here https://github.com/pytroll/satpy/blob/fa56be53f8f78660b76f95bde7d382cb6552edd9/satpy/readers/li_base_nc.py#L409
could you please check if they are not daskified here already, and if so, why? And otherwise, if they're always daskified coming from the file, we should find out where in the code they are computed and turned into numpy...

**Comment by ameraner** [comment]
Hi @ClementLaplace , thanks for the analysis! I see, it's quite as @gerritholl was suspecting :) Ok, then I would say it is ok to keep this implementation, in order to make sure that we have homogenised dask array outputs for dataset.

**Pull request merged**